src/parcsr_mv/par_csr_filter.c

```diff
@@ -157,27 +162,48 @@ hypre_ParCSRMatrixBlkFilterHost( hypre_ParCSRMatrix  *A,
       B_offd_i[i + 1] = B_offd_i[i];
       for (j = A_offd_i[i]; j < A_offd_i[i + 1]; j++)
       {
-         if (c == (HYPRE_Int) (col_map_offd_A[A_offd_j[j]] % big_block_size))
+         big_col = col_map_offd_A[A_offd_j[j]];
+         if (c == (HYPRE_Int) (big_col % big_block_size))
          {
-            B_offd_j[B_offd_i[i + 1]] = A_offd_j[j];
-            B_offd_a[B_offd_i[i + 1]] = A_offd_a[j];
+            B_offd_bj[B_offd_i[i + 1]] = big_col;
+            B_offd_a[B_offd_i[i + 1]]  = A_offd_a[j];
             B_offd_i[i + 1]++;
-            marker[A_offd_j[j]] = 1;
          }
       }
    }
 
-   /* Update col_map array */
-   num_cols_offd_B = 0;
-   for (i = 0; i < num_cols_offd_A; i++)
+   /* Allocate work array */
+   work = hypre_TAlloc(HYPRE_BigInt, B_offd_nnz, HYPRE_MEMORY_HOST);
```

Comment:
Why do you need an extra work array? Can you just copy directly into `col_map_offd_B` and work with that array below?

Reply:
I need the extra work array because the binary search below works on `B_offd_bj`

Reply:
just adding to the previous response: the work array is sorting out `B_offd_bj` in order to build col_map_offd. Unfortunately, we can't copy directly into col_map_offd without sorting first

Reply:
Sorry, I'm probably just missing something here... Why do you have to sort before copying into `col_map_offd_B`? It looks like `col_map_offd_B` is not used until after you copy the `work` array into it. So can't you just copy `B_offd_bj` into `col_map_offd_B`, then do the sorting and removal of duplicates in place in `col_map_offd_B`?

Reply:
No worries! I need to sort *and remove duplicate columns* before copying because the col_map_offd array gives a mapping for the compressed offd matrix. I can't copy `B_offd_bj` into `col_map_offd` because the first generally has more entries than the second one.

There's a similar code in the matrix assembly function: https://github.com/hypre-space/hypre/blob/79ab05654eacd34acd7c07e0a2d58f46008e0b43/src/IJ_mv/IJMatrix_petsc.c#L885-L914

Reply:
@waynemitchell not sure if my previous answer was clear. Please, let me know if I'm missing something

Reply:
@victorapm , I see it now. I had it in my head that `col_map_offd` was the same allocation size as these other arrays. Thanks for explaining!

Reply:
You're welcome, thanks for looking carefully at this!